### PR TITLE
[multibody] Require World frame and body frame for floating base bodies.

### PR DESCRIPTION
**Pull request by sherm1**
This PR tightens the definition of a "floating base body" to mean a body with a 6-dof joint connecting its _body frame_ directly to the _World frame_ (not just the World _body_). Previously it included 6-dof bodies where the parent frame was any frame on the World body, and the child frame was any frame on the free body. We believe that was an accident since the intent was that a "floating base body" should be one whose joint coordinates are World coordinates. That allows specifying the pose of a floating base body in World _prior_ to Finalize().

We've determined that none of Drake, Anzu, or Russ's course repos depended at all on the looser definition. We don't know about outside users but it is very unlikely anyone will notice this change. Still it is a potentially breaking change and should be noted as such in release notes.

This is a tiny code change but requires substantial updates to the documentation that goes with MultibodyPlant's "FreeBody" APIs. I've also tried to be very careful about the terminology since there are some tricky edge cases.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/23303)

### CODE REVIEW

**Comment by sherm1** [comment]
Results so far: all Drake tests pass.

**Comment by sherm1** [comment]
Drake has no problem with this change. If I understand the Anzu build correctly, it was also unaffected. The build was green in the Jenkins Anzu-with-Drake-PR-UI. In the Console log for build 305 I see:
```
13:59:56  commit 639bad4a48a1674aca354ae8d3b53c55d3bbebc1
13:59:56  Author: Michael Sherman <sherm1@gmail.com>
13:59:56  Date:   Thu Aug 21 15:54:15 2025 -0700
13:59:56  
13:59:56      Require World _frame_ (not just World _body_) for floating base bodies.
```
showing that it built with this commit. Then towards the end I see:
```
14:21:32  Executed 510 out of 7866 tests: 7866 tests pass.
...
14:21:54  Executed 2013 out of 2013 tests: 2013 tests pass.
```
I did see a timeout in the log.

Does this look like a clean Anzu build?

**Comment by jwnimmer-tri** [comment]
Yes, all good there.  LMK if you'd like me to test Russ's repositories?

@drake-jenkins-bot linux-noble-unprovisioned-gcc-wheel-experimental-release please

**Comment by sherm1** [comment]
> Yes, all good there. LMK if you'd like me to test Russ's repositories?

Yes, please. I'm not sure how to do that.

**Comment by jwnimmer-tri** [comment]
Both https://github.com/RussTedrake/underactuated and https://github.com/RussTedrake/manipulation pass `bazel test //...` when run locally against the experimental wheel built from this PR.  (Modulo tests that also fail with our stable release, e.g., broken html link checkers that gripe about websites that have gone missing.)

Here's the patch I used for testing.  This is the `manipulation` patch; the `underactuated` patch is similar.

```patch
--- a/requirements-bazel-linux.txt
+++ b/requirements-bazel-linux.txt
@@ -29,7 +29,6 @@ decorator==5.2.1 ; python_version >= "3.10" and python_version < "4.0"
 defusedxml==0.7.1 ; python_version >= "3.10" and python_version < "4.0" and (sys_platform == "darwin" or sys_platform == "linux")
 directsearch==1.0 ; python_version >= "3.10" and python_version < "4.0"
 distlib==0.3.9 ; python_version >= "3.10" and python_version < "4.0"
-drake==1.41.0 ; python_version >= "3.10" and python_version < "4.0"
 dulwich==0.21.7 ; python_version >= "3.10" and python_version < "4.0"
 embreex==2.17.7.post6 ; platform_machine == "x86_64" and python_version >= "3.10" and python_version < "4.0"
 entrypoints==0.4 ; python_version >= "3.10" and python_version < "4.0" and sys_platform == "linux"
@@ -174,3 +173,6 @@ widgetsnbextension==4.0.14 ; python_version >= "3.10" and python_version < "4.0"
 xxhash==3.5.0 ; python_version >= "3.10" and python_version < "4.0"
 yourdfpy==0.0.57 ; python_version >= "3.10" and python_version < "4.0"
 zipp==3.21.0 ; python_version >= "3.10" and python_version < "3.12"
+
+drake @ https://drake-packages.csail.mit.edu/drake/experimental/drake-0.0.20250825.215848%2Bgitde049857-cp312-cp312-manylinux_2_34_x86_64.whl
+Mosek==11.0.24
```

**Comment by sherm1** [comment]
Thanks Jeremy! I'll work this up into a real PR.

**Review by sherm1 — COMMENTED** [review]
+assignee:@SeanCurtis-TRI for feature review, please

This is ready for a first look, Sean. I'm particularly interested in your thoughts about the changes to the MultibodyPlant "FreeBody" API documentation and terminology. I have a few things to finish up but would like to get a second opinion on that stuff before I mangle things more.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-:-OYiRFRYEv0f8C_H-OWN:b-dgnnpc)* status: LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers

**Review by SeanCurtis-TRI — COMMENTED** [review]
FIrst pass done -- our f2f discussion is memoized here. I look forward to seeing the next version.

> @SeanCurtis-TRI reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-:-OYlpuf03q7cDLJdobZ6:blg6tqf)* status: 10 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers

___
*[`multibody/plant/multibody_plant.h` line 3329 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYlJ65C3Xc49omvthKn:-OYlJ65C3Xc49omvthKo:br6ozx8) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3329)):*
BTW The word "Node" jumped out at me oddly. Perhaps a tighter connection with the previous sentence would help:

> of the system as a forest of trees, in which bodies are represented as nodes and joints as edges.

Or some such thing.



_Code quote:_
```objc
Nodes 
```
___
*[`multibody/plant/multibody_plant.h` line 3336 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYlJqI588282pC8w4kG:-OYlJqI588282pC8w4kH:b28h59u) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3336)):*
> ```objc
>   /// joint used; typically it is a QuaternionFloatingJoint.) Thus there is
>   /// always a unique joint connecting a body to its parent in the "inboard"
>   /// direction (that is, closer to World), so for convenience below we may
> ```

nit, punctuation



_Suggestion:_
```objc
 so, for convenience below, we may
```
___
*[`multibody/plant/multibody_plant.h` line 3337 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYlJyo72Cic_sRQzZBK:-OYlJyo72Cic_sRQzZBL:b-36emmq) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3337)):*
> ```objc
>   /// always a unique joint connecting a body to its parent in the "inboard"
>   /// direction (that is, closer to World), so for convenience below we may
>   /// refer to Jc as the _body's_ child frame and Jp the _body's_ parent frame,
> ```

BTW I *really* don't like referring to Jc as the "body's child frame". Using this nomenclature, the body's parent frame is always fixed to a *different* body, but the body's child frame is always fixed to itself. Given the idea of a tree, this kind of language typically implies multiple layers of nodes (one layer up for parents and one layer down for children).

So, we are simultaneously using language associated with a well-known abstraction (trees, nodes and edges) but then using closely related language to mean something else. It seems to invite confusion.

-------------------------------------

Having read further, I've come up with an alternate word. We keep "body parent frame" but throw out "body child frame" in favor of "body inboard frame".
___
*[`multibody/plant/multibody_plant.h` line 3340 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYlMGEv2o0f6yGgFaaE:-OYlMGEv2o0f6yGgFaaF:b-geboro) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3340)):*
> ```objc
>   /// though these really belong to that body's unique inboard joint.
>   ///
>   /// In this representation each input joint (including those added for base
> ```

BTW I'm really not comfortable with the word "input" in this context. At the *plant* level they seem to be the joints and bodies that the user has added by calls to "Addjoint", "AddBody", etc.

But in this context "input" also means the implicit joints because nothing was specified. "input" is a good word for the tree, but it doesn't feel like a good word for the *plant*.

This predates *this* PR so you can ignore it. But it struck me hard, so I thought I'd at least share the trauma.



_Code quote:_
```objc
input joint
```
___
*[`multibody/plant/multibody_plant.h` line 3348 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYlMnrhEWY5ZMZf1FlT:-OYlMnrhEWY5ZMZf1FlU:b80b4hk) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3348)):*
> ```objc
>   /// call it a _floating base body_. A floating base body C has the useful
>   /// property that its child joint frame Jc can be posed arbitrarily using
>   /// joint coordinates that are also World coordinates. Most commonly, we also
> ```

nit: I'm not sure what "world coordinates" means. Joint coordinates, in this case, could include RPY values or quaternion values. Are they both equally "world coordinates"?

The language we'd bandied about the other day was that the pose of a floating base body in the world can be calculated from its joint coordinates directly, without consideration of any other body. I don't know that's the same as "the joint coordinates are also World coordinates".

Also, "can be posed arbitrarily" is true of *every* free body. I think what you really mean is "can be given an arbitrary pose in World". This is another aspect where I think this characterization misses the property I have in mind (described above).



_Code quote:_
```objc
 World coordinates
```
___
*[`multibody/plant/multibody_plant.h` line 3352 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYlNkgi9fREtnYut3FU:-OYlNkgi9fREtnYut3FV:b-51msa6) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3352)):*
> ```objc
>   /// floating base body in World with no need for frame conversions. That is
>   /// the case for the very common scenario where a body is added to the plant
>   /// without specifying a joint, so is automatically made a floating base body
> ```

nit precision



_Suggestion:_
```objc
without specifying its inboard joint
```
___
*[`multibody/plant/multibody_plant.h` line 3364 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYlO9oq8DQbl5ykQ1xr:-OYlO9oq8DQbl5ykQ1xs:b-al3mmd) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3364)):*
> ```objc
>   /// whether a RigidBody is a floating base body with RigidBody::is_floating().
>   /// Floating base bodies have the unique feature that, using the
>   /// SetDefaultFreeBodyPose() API below, you can set their default poses in
> ```

BTW I feel we should say something explicit about the fact that the *name* of this function is no longer "correct" (and hasn't been for a while). In the context of this paragraph the "set free body pose" function doesn't operate on free bodies and doesn't have the affect described here for all "floating base bodies" (particularly if Jc != C).

If we're adding this function name into *this* conversation, I think we need to head the inevitable confusion off at the pass.



_Code quote:_
```objc
SetDefaultFreeBodyPose()
```
___
*[`multibody/plant/multibody_plant.h` line 3432 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYljXQa176vY5cl9NOI:-OYljXQa176vY5cl9NOJ:b-t3umho) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3432)):*
> ```objc
>   /// @note The parent frame Jp is not necessarily the World frame W, and the
>   ///   child frame Jc is not necessarily the body frame C.
>   /// @see @ref mbp_working_with_free_bodies "above for details".
> ```

nit I don't know if you realize how this renders:



You might be better off *not* using the `@see` tag and just doing:

`see @ref mbp_working_with_free_bodies "above for details".`

at the end of the documentation.



_Code quote:_
```objc
@see @ref mbp_working_with_free_bodies "above for details".
```
___
*[`multibody/plant/multibody_plant.h` line 3492 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYlkkjB1eJWe8Pg351y:-OYlkkjB1eJWe8Pg351z:bfus74n) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3492)):*
> ```objc
>   ///   RigidBody whose default pose will be retrieved.
>   /// @retval X_WC The default pose of the floating base body C in World. Not
>   ///   meaningful if `body` is not a floating base body.
> ```

Btw you have the "meaningful" warning *twice* in one function's documentation. On the return value and an explicit "warning". Overkill?
___
*[`multibody/plant/multibody_plant.h` line 3624 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/23303#-OYllRfA7wQ-UNRngMJQ:-OYllRfA7wQ-UNRngMJR:bunc6z7) ([raw file](https://github.com/RobotLocomotion/drake/blob/9c876371d7562f2420fb4772acba5a4369d63f03/multibody/plant/multibody_plant.h#L3624)):*
> ```objc
>                                    const math::RigidTransform<T>& X_WJc) const;
> 
>   /// Updates `context` to store the pose `X_FJc` of a given floating base body
> ```

BTW I find while this documentation is more precise, it seems to illuminate a defect in the API.

If we're passing a body and a pose, you'd like to think you're posing the *body*. The fact that it has an inboard frame that isn't coincident with its body frame should contribute to the calculations, but it should still be the pose of the body.

**Comment by sherm1** [comment]
Per f2f with Sean I'm going to try further restricting "floating base body" to require the _child_ frame to be the body frame, which is also likely to be the case in existing code, is much more useful, and will simplify the documentation. I've got to rerun build tests again to make sure this doesn't break anything, so will switch back to Draft/don't review for a while.

**Comment by sherm1** [comment]
This Draft PR has served its purpose. Will replace with a real PR shortly that will address comments from here and #23349 (also closed).
